packages/agreement-process/src/services/readModelService.ts

```diff
@@ -318,51 +319,108 @@ async function searchTenantsByName(
+                      ? "$$consumerId"
+                      : "$$producerId",
+                  ],
+                },
+                {
+                  $eq: ["$data.kind", kind],
+                },
+                {
+                  $eq: ["$data.state", delegationState.active],
+                },
+              ],
+            },
           },
         },
-        {
-          $unwind: {
-            path: "$delegations",
-            preserveNullAndEmptyArrays: true,
-          },
+      ],
+      as: "delegations",
+    },
+  },
+];
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+function getProducerOrConsumerWithDelegateFilters(
+  producerIds: TenantId[],
+  consumerIds: TenantId[],
+  requesterId: TenantId
+) {
+  const requesterRole: "Consumer" | "Producer" | undefined =
+    producerIds.length === 1 && producerIds[0] === requesterId
+      ? "Producer"
+      : consumerIds.length === 1 && consumerIds[0] === requesterId
+      ? "Consumer"
+      : undefined;
+
+  return match([requesterRole, producerIds.length > 0, consumerIds.length > 0])
+    .with(["Producer", false, P.any], () => []) // Impossible case (producerIds must be set)
+    .with(["Consumer", P.any, false], () => []) // Impossible case (consumerIds must be set)
+    .with([undefined, P.any, P.any], () => [
+      {
+        $match: {
+          ...makeFilter("producerId", producerIds),
+          ...makeFilter("consumerId", consumerIds),
         },
-        {
-          $match: {
-            $or: [
-              {
-                $and: [
-                  {
-                    "delegations.data.kind": delegationKind.delegatedProducer,
-                  },
-                  {
-                    "delegations.data.state": agreementState.active,
-                  },
-                  {
-                    "delegations.data.delegateId": {
-                      $in: producerIds,
-                    },
-                  },
-                ],
-              },
-              {
-                "data.producerId": {
-                  $in: producerIds,
+      },
+    ])
+    .with(["Producer", true, P.any], () => [
+      ...getDelegationLookup(delegationKind.delegatedProducer),
+      {
+        $match: {
+          $and: [
+            {
+              $or: [
+                {
+                  "delegations.data.delegateId": requesterId,
                 },
-              },
-            ],
-          },
+                makeFilter("producerId", requesterId),
+              ],
+            },
+            makeFilter("consumerId", consumerIds),
+          ].filter((a) => a !== undefined),
+        },
+      },
+    ])
+    .with(["Consumer", P.any, true], () => [
+      ...getDelegationLookup(delegationKind.delegatedConsumer),
+      {
+        $match: {
+          $and: [
+            {
+              $or: [
+                {
+                  "delegations.data.delegateId": requesterId,
+                },
```

Comment:
Since this is the same here and above, it could probably be moved inside getDelegationLookup? Adding the requesterId as a second parameter after the kind

Reply:
I don't think it is possible. In the lookup pipeline I'm doing the "join" but here I want to filter out data. Besides, I would complicate the logic of that method since if you are consumer you need to filter, potentially, also for producer and vice-versa.